### PR TITLE
CPU diagonal assembly for multiple active bases

### PR DESCRIPTION
**Pull request by jeremylt**
I need to fix up all diagonal assembly to work like this and make the fully assembly when num_active_bases > 1 for this initial PR.

### CODE REVIEW

**Comment by jeremylt** [comment]
The test I added works, so @rezgarshakeri you should be able to test Jacobi based PCs with this branch for your mixed elasticity Ratel MR

**Comment by jeremylt** [comment]
As a heads up @nbeams, these updates need to be ported to the GPU backends in another PR and then the full assembly needs to be updated in another couple of PRs. I don't have a timeline for that yet.

**Comment by nbeams** [comment]
> As a heads up @nbeams, these updates need to be ported to the GPU backends in another PR and then the full assembly needs to be updated in another couple of PRs. I don't have a timeline for that yet.

Let me know at what point (if any) you want me to hop in and work on GPU stuff.  I was assuming maybe at the full assembly point, since that was my code originally -- is that what your comment is referring to?

**Comment by jeremylt** [comment]
I don't have a plan of attack yet, I more just wanted to put it on your radar since it will involve code you wrote. 

I think adding this diagonal assembly change to the GPU should be easier but the full assembly update will be harder.

**Comment by rezgarshakeri** [comment]
> The test I added works, so @rezgarshakeri you should be able to test Jacobi based PCs with this branch for your mixed elasticity Ratel MR

Thanks @jeremylt, I'll try it.

File: interface/ceed-preconditioning.c [hunk 1]
```diff
@@ -1160,132 +1262,172 @@ int CeedOperatorAssemblyDataCreate(Ceed ceed, CeedOperator op, CeedOperatorAssem
...
+  @param[out] num_active_bases      Number of active bases
+  @param[out] num_eval_mode_in      Pointer to hold array of numbers of input CeedEvalModes, or NULL
+  @param[out] eval_mode_in          Pointer to hold arrays of input CeedEvalModes, or NULL
+  @param[out] eval_mode_offsets_in  Pointer to hold arrays of input offsets at each quadrature point
+  @param[out] num_eval_mode_out     Pointer to hold array of numbers of output CeedEvalModes, or NULL
+  @param[out] eval_mode_out         Pointer to hold arrays of output CeedEvalModes, or NULL
+  @param[out] eval_mode_offsets_out Pointer to hold arrays of output offsets at each quadrature point
+  @param[out] num_output_components Total number of output components at each quadrature point
 
   @return An error code: 0 - success, otherwise - failure
```
Comment (jedbrown) [inline interface/ceed-preconditioning.c:1287]: I think this man page would benefit from an explanation of the data model. Does `num_active_bases` apply to in or out? How is `num_output_components` counted across the various fields?

**Review by jedbrown — APPROVED** [review]

File: interface/ceed-preconditioning.c (hunk 1 above)
Comment (jeremylt) [reply interface/ceed-preconditioning.c:1287]: Agreed - the names are less obvious in the light of Monday morning.

File: interface/ceed-preconditioning.c [hunk 2]
```diff
@@ -569,13 +573,20 @@ static int CeedSingleOperatorAssemble(CeedOperator op, CeedInt offset, CeedVecto
...
+                                                &eval_modes_out, NULL, NULL));
+  CeedBasis *bases;
+  CeedCall(CeedOperatorAssemblyDataGetBases(data, NULL, &bases, NULL, NULL));
+  CeedBasis basis_in = bases[0];
+
+  if (num_active_bases > 1) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_UNSUPPORTED, "Cannot assemble operator with multiple active bases");
+    // LCOV_EXCL_STOP
+  }
```
Comment (rezgarshakeri) [inline interface/ceed-preconditioning.c:588]: This means we still don't support full assembly for multiple bases, right?

Comment (jeremylt) [reply interface/ceed-preconditioning.c:588]: Correct. That will be much, much more work

**Review by rezgarshakeri — APPROVED** [review]

**Pull request merged**